caffe2/python/caffe_translator.py

```diff
@@ -92,36 +92,50 @@ def TranslateModel(
                 'only accepts new style layers that are stored in the '
                 'layer field.'
             )
-        for layer in caffe_net.layer:
-            if not _ShouldInclude(net_state, layer):
+        i = 0
+        pretrained_layers = [l for l in pretrained_net.layer] +\
+                            [l for l in pretrained_net.layers]
+        while i < len(caffe_net.layer):
+            if not _ShouldInclude(net_state, caffe_net.layer[i]):
                 log.info('Current net state does not need layer {}'
-                            .format(layer.name))
+                            .format(caffe_net.layer[i].name))
                 continue
-            log.info('Translate layer {}'.format(layer.name))
+            log.info('Translate layer {}'.format(caffe_net.layer[i].name))
             # Get pretrained one
-            pretrained_layers = (
-                [l for l in pretrained_net.layer
-                 if l.name == layer.name] + [l
-                                             for l in pretrained_net.layers
-                                             if l.name == layer.name]
-            )
-            if len(pretrained_layers) > 1:
+            layers_index = [l for l in xrange(len(pretrained_layers))
+                 if pretrained_layers[l].name == caffe_net.layer[i].name]
+            is_bn = False
+            if len(layers_index) > 1:
                 raise ValueError(
                     'huh? more than one pretrained layer of one name?')
-            elif len(pretrained_layers) == 1:
-                pretrained_blobs = [
-                    utils.CaffeBlobToNumpyArray(blob)
-                    for blob in pretrained_layers[0].blobs
-                ]
+            elif len(layers_index) == 1:
+                if pretrained_layers[layers_index[0]].type == "BatchNorm":
+                    # A Scale layer should follow BatchNorm layer
+                    # according to paper https://arxiv.org/abs/1502.03167.
+                    assert pretrained_layers[layers_index[0]+1].type == "Scale"
```

Comment:
Actually, this is correct. In Caffe the BatchNorm layer can't handle a bias, where as in Caffe2, the bias is required for the SpatialBN operator. Also, since Caffe2's Scale operator can't handle a bias we can only do the conversion if a Scale layer with a bias term follows the batch norm layer.